file_store.go

```diff
@@ -0,0 +1,275 @@
+
+	if _, err := os.Stat(fs.configPath); os.IsNotExist(err) {
+		// no ops if the config file does not exist yet
+		return nil
+	}
+	if _, ok := fs.authsCache[serverAddress]; !ok {
+		// no ops
+		return nil
+	}
+
+	delete(fs.authsCache, serverAddress)
+	return fs.saveFile()
+}
+
+// saveFile saves fs.content into fs.configPath.
+func (fs *FileStore) saveFile() error {
+	if len(fs.authsCache) > 0 {
+		authsBytes, err := json.Marshal(fs.authsCache)
+		if err != nil {
+			return fmt.Errorf("failed to marshal credentials: %w", err)
+		}
+		fs.content[configFieldAuths] = authsBytes
+	} else {
+		// omit empty
+		delete(fs.content, configFieldAuths)
+	}
+	jsonBytes, err := json.MarshalIndent(fs.content, "", "\t")
+	if err != nil {
+		return fmt.Errorf("failed to marshal config: %w", err)
+	}
+
+	configDir := filepath.Dir(fs.configPath)
+	if err := os.MkdirAll(configDir, 0700); err != nil {
+		return fmt.Errorf("failed to make directory %s: %w", configDir, err)
+	}
+	ingest, err := ioutil.Ingest(configDir, bytes.NewReader(jsonBytes))
+	if err != nil {
+		return fmt.Errorf("failed to save config file: %w", err)
+	}
+
+	// handle symlink
+	targetPath := fs.configPath
+	if link, err := os.Readlink(fs.configPath); err == nil {
+		targetPath = link
+	}
+	// copy file with original ownership and permissions
+	ioutil.CopyFilePermissions(targetPath, ingest)
+	if err := os.Rename(ingest, targetPath); err != nil {
+		// clean up the ingest file
+		os.Remove(ingest)
+		return fmt.Errorf("failed to save config file: %w", err)
+	}
+	return nil
+}
+
+// authConfigToCredential converts an authConfig to an auth.Credential.
+func authConfigToCredential(authCfg authConfig) (auth.Credential, error) {
+	cred := auth.Credential{
+		Username:     authCfg.Username,
+		Password:     authCfg.Password,
+		RefreshToken: authCfg.IdentityToken,
+		AccessToken:  authCfg.RegistryToken,
+	}
+	if authCfg.Auth != "" {
+		var err error
+		// override username and password
+		cred.Username, cred.Password, err = decodeAuth(authCfg.Auth)
+		if err != nil {
+			return auth.EmptyCredential, fmt.Errorf("failed to decode auth field: %w: %v", ErrInvalidConfigFormat, err)
+		}
+	}
+	return cred, nil
+}
+
+// credentialToAuthConfig converts an auth.Credential to an authConfig.
+func credentialToAuthConfig(cred auth.Credential) authConfig {
+	return authConfig{
+		Auth:          encodeAuth(cred.Username, cred.Password),
+		IdentityToken: cred.RefreshToken,
+		RegistryToken: cred.AccessToken,
+	}
+}
+
+// encodeAuth base64-encodes username and password into base64(username:password).
+// Reference: https://github.com/docker/cli/blob/v24.0.0-beta.1/cli/config/configfile/file.go#L215-L226
+func encodeAuth(username, password string) string {
+	if username == "" && password == "" {
+		return ""
+	}
+
+	authStr := username + ":" + password
+	msg := []byte(authStr)
+	encoded := make([]byte, base64.StdEncoding.EncodedLen(len(msg)))
+	base64.StdEncoding.Encode(encoded, msg)
+	return string(encoded)
```

Comment:
You may use a method named [EncodeToString()](https://pkg.go.dev/encoding/base64#Encoding.EncodeToString).